test/e2e/k8s_util.go

```diff
@@ -724,34 +726,47 @@ func (k *KubernetesUtils) waitForHTTPServers(allPods []Pod) error {
 	return errors.Errorf("after %d tries, %d HTTP servers are not ready", maxTries, wrong)
 }
 
-func (k *KubernetesUtils) Validate(allPods []Pod, reachability *Reachability, port int32, protocol v1.Protocol) {
+// Validate checks the connectivity between all Pods in both direction with a
+// list of ports and a protocol. The connectivity from a Pod to another Pod should
+// be consistent on all provided ports. Otherwise, this connectivity will be
+// treated as Error
+func (k *KubernetesUtils) Validate(allPods []Pod, reachability *Reachability, ports []int32, protocol v1.Protocol) {
 	type probeResult struct {
 		podFrom      Pod
 		podTo        Pod
 		connectivity PodConnectivityMark
 		err          error
 	}
-	numProbes := len(allPods) * len(allPods)
+	numProbes := len(allPods) * len(allPods) * len(ports)
 	resultsCh := make(chan *probeResult, numProbes)
 	// TODO: find better metrics, this is only for POC.
-	oneProbe := func(podFrom, podTo Pod) {
+	oneProbe := func(podFrom, podTo Pod, port int32) {
 		log.Tracef("Probing: %s -> %s", podFrom, podTo)
 		connectivity, err := k.Probe(podFrom.Namespace(), podFrom.PodName(), podTo.Namespace(), podTo.PodName(), port, protocol)
 		resultsCh <- &probeResult{podFrom, podTo, connectivity, err}
 	}
 	for _, pod1 := range allPods {
 		for _, pod2 := range allPods {
-			go oneProbe(pod1, pod2)
+			for _, port := range ports {
+				go oneProbe(pod1, pod2, port)
+			}
 		}
 	}
 	for i := 0; i < numProbes; i++ {
 		r := <-resultsCh
 		if r.err != nil {
 			log.Errorf("unable to perform probe %s -> %s: %v", r.podFrom, r.podTo, r.err)
 		}
-		reachability.Observe(r.podFrom, r.podTo, r.connectivity)
+
+		preConn := reachability.Observed.Get(r.podFrom.String(), r.podTo.String())
```

Comment:
nit: trying to understand what `precConn` stands for here

Reply:
Previous Connectivity. Changed to `prevConn`